hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/CryptoGetAccountBalanceHandler.java

```diff
@@ -82,19 +86,47 @@ public void validate(@NonNull final QueryContext context) throws PreCheckExcepti
         final var query = context.query();
         final var accountStore = context.createStore(ReadableAccountStore.class);
         final CryptoGetAccountBalanceQuery op = query.cryptogetAccountBalanceOrThrow();
+        mustExist(op.balanceSource(), INVALID_CONTRACT_ID);
+        if (op.balanceSource().kind().equals(UNSET)) {
+            throw new PreCheckException(INVALID_ACCOUNT_ID);
+        }
         if (op.hasAccountID()) {
-            final var account = accountStore.getAliasedAccountById(requireNonNull(op.accountID()));
-            validateFalsePreCheck(account == null, INVALID_ACCOUNT_ID);
-            validateFalsePreCheck(account.deleted(), ACCOUNT_DELETED);
+            validateAccountId(op, accountStore);
         } else if (op.hasContractID()) {
-            final var contract = accountStore.getContractById(requireNonNull(op.contractID()));
-            validateFalsePreCheck(contract == null || !contract.smartContract(), INVALID_CONTRACT_ID);
-            validateFalsePreCheck(contract.deleted(), CONTRACT_DELETED);
+            validateContractId(op, accountStore);
         } else {
             throw new PreCheckException(INVALID_ACCOUNT_ID);
         }
     }
 
+    private void validateContractId(CryptoGetAccountBalanceQuery op, ReadableAccountStore accountStore)
+            throws PreCheckException {
+        mustExist(op.contractID(), INVALID_CONTRACT_ID);
+        ContractID contractId = (ContractID) op.balanceSource().value();
+        validateTruePreCheck(contractId.shardNum() == 0, INVALID_CONTRACT_ID);
+        validateTruePreCheck(contractId.realmNum() == 0, INVALID_CONTRACT_ID);
+        validateTruePreCheck(contractId.contractNumOrThrow() >= 0, INVALID_CONTRACT_ID);
+        final var contract = accountStore.getContractById(requireNonNull(op.contractID()));
+        validateFalsePreCheck(contract == null || !contract.smartContract(), INVALID_CONTRACT_ID);
```

Comment:
Personally I find calling this with compound boolean expressions like so – `validateFalse(<condition 1> || <condition2> etc.)` – very hard to read, since it requires negating the expression with boolean algebra in order to decipher the code intent. 

```
validateFalsePreCheck(contract == null, INVALID_CONTRACT_ID);
validateTruePreCheck(contract.smartContract(), INVALID_CONTRACT_ID);
```

Just a thought though, nothing inherently incorrect about this afaict

Reply:
Makes sense 👍